### PR TITLE
Temporary computation buffer

### LINKED ISSUE #1275: Framework for temporary computation buffers
# Feature request:

It is becoming increasingly common to use aggregated buffers to store temporary variables. This is important to performance, as it minimizes the number of allocations required, and allocations are a significant cost in highly parallel codes.

So far, similar objects have been used in #1177, #1179, #1161.

The design of these buffers has generated some discussion on its own, so this issue is to create a space for fleshing out those ideas.

A quick summary of points that have been made:
- These buffers are very similar to `Variables`, and in the case of containing vector types, they are a thin wrapper for `Variables`. However, `Variables` cannot be used directly because it cannot (and probably should not be modified to) contain `Tensor`s of doubles.
- The additional wrapping permits more descriptive names to be assigned to the buffers. This can be both a positive and negative structure, as there are times where a single buffer entry may be used for the storage of multiple, unrelated, pieces of data in different parts of a computation.
- The proliferation of specialized buffers will tend to create a large number of data types which are only well-applicable to individual computations. This is a problem solved by the structure of tags and `Variables`, and the problem is far more mild in the case of the thin wrapping of the buffers, but it still seems undesirable to have all of these data types if a more generic data type is possible. 

### Component:

- [ ] Code

### PR DESCRIPTION
**Pull request by markscheel**
## Proposed changes

Adds TempBuffer, a buffer to store temporary variables in a way to reduce allocations.

Resolves #1275

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/DataStructures/TempBuffer.hpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ */
+template <typename TagList,
+          bool is_fundamental = cpp17::is_fundamental_v<
+              typename tmpl::front<TagList>::type::value_type>>
+struct TempBuffer;
+
+template <typename TagList>
+struct TempBuffer<TagList, true> : tuples::tagged_tuple_from_typelist<TagList> {
+  explicit TempBuffer(const size_t /*size*/) noexcept
+      : tuples::tagged_tuple_from_typelist<TagList>::TaggedTuple(){};
```
Comment (kidder) [inline src/DataStructures/TempBuffer.hpp]: extra ;

**Review by kidder — CHANGES_REQUESTED** [review]
group review with @nilsdeppe and @wthrowe

**Review by moxcodes — DISMISSED** [review]

**Review by nilsvu — DISMISSED** [review]

**Comment by markscheel** [comment]
Does anyone understand why IWYU wants me to include `<complex>` in KerrSchild.cpp?

**Comment by nilsdeppe** [comment]
It's confused and thinks the math functions like `sin`, etc. are in `<complex>` (which they are, but only for complex numbers...)

**Comment by markscheel** [comment]
Removed semicolon and added pragmas to make IWYU happy; squashed since changes are trivial.

**Review by moxcodes — DISMISSED** [review]

**Review by nilsdeppe — DISMISSED** [review]

**Review by kidder — DISMISSED** [review]

**Comment by wthrowe** [comment]
You squashed incorrectly.  I'll let the rest of you all decide whether you want it fixed or you don't care.

**Comment by markscheel** [comment]
I fixed my squash (the first time, I squashed a fixup into the wrong commit); it was easy to fix.

**Review by moxcodes — DISMISSED** [review]

**Review by wthrowe — DISMISSED** [review]

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@markscheel this has a conflict

**Comment by markscheel** [comment]
rebased on develop, no more conflicts.

**Review by nilsdeppe — APPROVED** [review]

**Review by moxcodes — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Review by moxcodes — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]

**Comment by nilsvu** [comment]
That's a lot of approvals for 64 lines of added code

**Comment by markscheel** [comment]
The failing test, a gcc debug test, actually passes all the tests but exceeds the time limit for some reason.

**Comment by kidder** [comment]
travis failure is timeout after tests passed

**Pull request merged**